lib/keyboard_utils.coffee

```diff
@@ -16,17 +16,31 @@ KeyboardUtils =
       @platform = "Windows"
 
   getKeyChar: (event) ->
-    if event.key of @keyNames
-      @keyNames[event.key]
-    # It appears that event.key is not always defined (see #2453).
-    else if not event.key?
+    unless Settings.get "ignoreKeyboardLayout"
+      key = event.key
+    else if event.code[...6] == "Numpad"
+      # We cannot correctly emulate the numpad, so fall back to event.key; see #2626.
+      key = event.key
+    else
+      # The logic here is from the vim-like-key-notation project (https://github.com/lydell/vim-like-key-notation).
+      key = event.code
+      key = key[3..] if key[...3] == "Key"
+      # Translate some special keys to event.key-like strings.
+      if @enUsTranslations[key]
+        key = if event.shift then @enUsTranslations[key][1] else @enUsTranslations[key][0]
+      key = key.toLowerCase() unless event.shiftKey
```

Comment:
Hmm ... won't this result in inconsistent results for non-char-keys?

- <kbd>Enter</kbd> → `enter`
- <kbd>Shift</kbd><kbd>Enter</kbd> → `Enter`